### PR TITLE
First attempt at linter

### PR DESCRIPTION
**Pull request by gsnedders**
~This probably needs work. And might not work. And should probably have some Travis script first. (i.e., don't merge it)~

Probably good enough.

### CODE REVIEW

**Comment by gsnedders** [comment]
I don't know if the only reason to vendor `funcpaserlib` was https://github.com/vlasovskikh/funcparserlib/issues/44, or whether this was to avoid external deps.

File: lint_lib/parser.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+from ._vendor.funcparserlib.lexer import Token, LexerError
+from ._vendor.funcparserlib.parser import (Parser, State, NoParseError,
+                                           finished, many, pure, skip, some)
+
+text_type = type("")
+binary_type = type(b"")
+
+
+def _make_tokenizer(specs):
+    # Forked from upstream funcparserlib.lexer to fix #44 and #46
```
Comment (gsnedders) [inline lint_lib/parser.py:14]: c.f. https://github.com/vlasovskikh/funcparserlib/issues/44 (won't fix) and https://github.com/vlasovskikh/funcparserlib/issues/46 (still open)

**Comment by gsnedders** [comment]
> I don't know if the only reason to vendor `funcpaserlib` was [vlasovskikh/funcparserlib#44](https://github.com/vlasovskikh/funcparserlib/issues/44), or whether this was to avoid external deps.

Pretty sure it was avoiding external deps.

**Comment by gsnedders** [comment]
FWIW: if nobody reviews this and nobody objects, I'm just going to land this in a week so we avoid regressing these things yet again.

File: lint_lib/lint.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        ],
+        input_headers={
+            "data",
+            "document-fragment",
+            "script-on",
+            "script-off",
+        },
+    )
+    if not parsed:
+        return
```
Comment (Ms2ger) [inline lint_lib/lint.py:250]: If this does something, it should do it in a less cryptic way

File: lint_lib/parser.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    _Tuple,
+    finished,
+    many,
+    pure,
+    skip,
+    some,
+    tok,
+)
+
+text_type = str
```
Comment (Ms2ger) [inline lint_lib/parser.py:17]: Can we search & replace this?

File: lint_lib/parser.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+def _make_tokenizer(specs: List[Tuple[str, Tuple[StringLike]]]) -> Callable:
+    # Forked from upstream funcparserlib.lexer to fix #46
+    def compile_spec(spec):
+        name, args = spec
+        return name, re.compile(*args)
+
+    compiled = [compile_spec(s) for s in specs]
+
+    def match_specs(specs, str, i, position):
+        if isinstance(str, text_type):
```
Comment (Ms2ger) [inline lint_lib/parser.py:40]: (I guess this argument would need another name)

**Review by Ms2ger — APPROVED** [review]

File: lint_lib/lint.py (hunk 2 above)
Comment (gsnedders) [reply lint_lib/lint.py]: @Ms2ger PTAL at the comments that are there now

File: lint_lib/lint.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+            "script-on",
+            "script-off",
+        },
+    )
+    if not parsed:
+        # We'll already have output if there's a parse error.
+        return
+
+    # We'd put extra linting here, if we ever have anything specific to the
+    # encoding tests here.
```
Comment (gsnedders) [inline lint_lib/lint.py:258]:
```suggestion
    # tree construction tests here.
```

**Review by Ms2ger — APPROVED** [review]
wfm

**Pull request merged**